[DNM] rgw: Adding code for Role creation, deletion and reading.

pritha-srivastava opened the pull request:
A new class RGWRoles for AWS roles has been added.
Also, code for role create. delete and list for radosgw-admin has been added.

Signed-off-by: Pritha Srivastava prsrivas@redhat.com

cbodley commented on src/rgw/rgw_role.cc:
```diff
@@ -0,0 +1,220 @@
+
+  // Creation time
+  creation_date = boost::posix_time::to_iso_extended_string(
+                          boost::posix_time::microsec_clock::universal_time());
```
does the standard library support this string format? i'd prefer to use `ceph::real_clock` instead of boost if at all possible

cbodley commented on src/rgw/rgw_admin.cc:
```diff
@@ -3955,7 +3986,50 @@ int main(int argc, char **argv)
+      }
+      show_role_info(role, formatter);
+      return 0;
+    }
```
the 'role list' command feels more like a 'role get', doesn't it? will there ever be multiple entries to output here?

cbodley commented on src/rgw/rgw_role.cc:
```diff
@@ -0,0 +1,220 @@
+  }
+
+  return 0;
+}
```
this function is just reading the role into memory, right? can we name it `init()` or `read()` instead?

cbodley commented on src/rgw/rgw_role.cc:
```diff
@@ -0,0 +1,220 @@
+
+int RGWRole::read_name()
+{
+  string pool_name = store->get_zone_params().roles_pool.name;
```
shouldn't need to make a copy of the string here. consider changing this to `auto& pool = store->get_zone_params().roles_pool`, so it be used for both the `rgw_get_system_obj()` call, and to access the name as `pool.name`. can do similar in the other functions too

pritha-srivastava commented:
@cbodley : I didn't add a separate oid for role names because, if a user tries to create a role with a name same as an existing role id, then the method create will return an error saying that the name is in use as an id by another role.
And I think allowing roles with the same name and id is slightly confusing (though with a different prefix). I know the convention has been to separate out ids and names. If we have to strictly follow that then I'll add oid prefixing for the time being till we switch over to rados namespacing. Let me know what you think.

pritha-srivastava commented:
I gave it some more thought, i guess it will be better to separate out the namespaces for roles, names etc (oid prefixes) for the time being, if going forward we would want to list the role names. For e.g. in case of list roles, we may have to add a separate namespace for path in order to list all roles under the same path.

cbodley commented on src/rgw/rgw_role.h:
```diff
@@ -0,0 +1,81 @@
+  const string& get_id() { return id; }
+  const string& get_name() { return name; }
+  const string& get_path() { return path; }
+  const string& get_create_date() { return creation_date; }
```
these functions can be `const`

cbodley commented on src/rgw/rgw_admin.cc:
```diff
@@ -158,6 +159,9 @@ void _usage()
   cout << "  orphans list-jobs          list the current job-ids for orphans search\n";
+  cout << "  role create                create a role\n";
+  cout << "  role delete                delete a role\n";
+  cout << "  role get                   get a role\n";
```
can we add a bit more context? something like `create an AWS role for use with STS`? `delete a role` and `get a role` are fine as they are

cbodley commented on src/rgw/rgw_admin.cc:
```diff
@@ -3955,7 +3983,50 @@ int main(int argc, char **argv)
+      RGWRole role(g_ceph_context, store, role_name, path);
+      ret = role.create(true);
+      if (ret < 0) {
+        cerr <<"could not create role: " << err_msg << std::endl;
```
nit: space after <<

also, where is `err_msg` coming from?

pritha-srivastava commented on src/rgw/rgw_rados.h:
```diff
@@ -937,7 +938,7 @@ struct RGWZoneParams : RGWSystemMetaObj {
   
   void encode(bufferlist& bl) const {
-    ENCODE_START(8, 1, bl);
+    ENCODE_START(9, 1, bl);
```
@cbodley : There was a merge conflict here; tier_config was added in master at version number 8, so i have bumped up the version again to 9, to add roles_pool. Is that correct?

cbodley replied:
:+1:

cbodley commented on src/rgw/rgw_admin.cc:
```diff
@@ -157,6 +158,9 @@ void _usage()
   cout << "  orphans list-jobs          list the current job-ids for orphans search\n";
+  cout << "  role create                create a AWS role for use with STS \n";
+  cout << "  role delete                delete a role\n";
+  cout << "  role get                   get a role\n";
```
nit: extra space after `STS`

also, there's a unit test that compares the output of `radosgw-admin -h` against `test/cli/radosgw-admin/help.t` - could you please update that file to match your changes? jenkins is complaining about it:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t: failed
--- /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t
+++ /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t.err
@@ -108,6 +108,9 @@
     orphans find               init and run search for leaked rados objects (use job-id, pool)
     orphans finish             clean up search for leaked rados objects
     orphans list-jobs          list the current job-ids for orphans search
+    role create                create a AWS role for use with STS 
+    role delete                delete a role
+    role get                   get a role
   options:
```

cbodley commented on src/rgw/rgw_admin.cc:
```diff
@@ -2416,6 +2440,10 @@ int main(int argc, char **argv)
+    } else if (ceph_argparse_witharg(args, i, &val, "--role-name", (char*)NULL)) {
+      role_name = val;
+    } else if (ceph_argparse_witharg(args, i, &val, "--path", (char*)NULL)) {
+      path = val;
```
these options need to be documented in the usage also

cbodley commented on src/rgw/rgw_admin.cc:
```diff
@@ -4017,7 +4045,50 @@ int main(int argc, char **argv)
+      if (role_name.empty()) {
+        cerr << "Empty role name" << std::endl;
+        return -EINVAL;
+      }
```
is `path` allowed to be empty, or should we check that here too?

pritha-srivastava replied:
@cbodley : path can be empty since it is optional. It is set to "/" if the user doesn't set it to anything (according to AWS create role definition.)

cbodley commented:
@pritha-srivastava looks good! i'm happy to merge once the `[DNM]` tag goes away

pritha-srivastava commented:
@cbodley : Removed the 'dnm' tag.

cbodley commented on src/rgw/rgw_admin.cc:
```diff
@@ -4017,7 +4047,50 @@ int main(int argc, char **argv)
+      if (ret < 0) {
+        cerr << "could not delete role: " << cpp_strerror(-ret) << std::endl;
+        return -ret;
+      }
```
could you add a `cout` here so the user knows it succeeded?

cbodley commented on src/rgw/rgw_role.cc:
```diff
@@ -0,0 +1,245 @@
+  ret = read_info();
+  if (ret < 0) {
+    ldout(cct, 0) << "ERROR: finding role id in pool: " << pool.name << ": "
+                  << id << ": " << cpp_strerror(-ret) << dendl;
```
both `read_name()` and `read_info()` are already logging error messages, so i don't think we need to log it again here:
```
$ radosgw-admin role get --role-name=foo
2016-10-18 09:53:41.912480 7f959a04ba80  0 ERROR: failed reading role name from pool: default.rgw.roles: foo: (2) No such file or directory
could not get role: (2) No such file or directory
2016-10-18 09:53:41.912580 7f959a04ba80  0 ERROR: finding role name in pool: default.rgw.roles: foo: (2) No such file or directory
```

pritha-srivastava commented:
@cbodley: Made the following changes:
1. Code clean up wrt usability.
2. Added error handling for the create when secondary index creation fails.
3. Added code for parsing trust policy document (json) and encoding that as a string for the time being.

pritha-srivastava commented on src/rgw/rgw_role.cc:
```diff
@@ -0,0 +1,275 @@
+                  << id << ": " << cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+}
```
@cbodley: Yeah initially, i thought of breaking the logic as you have suggested, but if we implement the AWS API, then the same code (that will be there in rgw-admin.cc), will be replicated in the class which will have the corresponding method for role modify API. Thats the reason why I put everything under one interface in RGWRole class. Let me know what you think.

cbodley replied:
ok, makes sense

mattbenjamin commented:
@adamemerson would be good to get your review here

pritha-srivastava commented on src/rgw/rgw_role.cc:
```diff
@@ -82,10 +84,10 @@ int RGWRole::create(bool exclusive)
+  oss << std::put_time(std::gmtime(&tv.tv_sec), "%Y-%m-%dT%H:%M:%S");
+  oss << "." << (tv.tv_usec/1000) << "Z";
+  creation_date = oss.str();
 
```
@mattbenjamin : Does the above look better as compared to strftime? I came across put_time which is there since C++11.

adamemerson replied:
C++ time formatting is, unfortunately, a bit of a fiasco so I'm not sure there's a really *good* solution. strftime is probably faster since it has less overhead, not needing to call the entire iostream infrastructure into place.

The one thing is, don't use std::gmtime. It's not guaranteed to be thread safe (and strict interpretation of the C standard forbids it to be thread safe, since all calls to gmtime and localtime are required to return the same pointer). Use gmtime_r. It's not C++ but is in the standard C library. It started out as a GNU extension but has been folded into the SIngle Unix Specification so should be available everywhere.

adamemerson commented:
This seems a fine PR to me, at least once we switch gmtime for gmtime_r.

Labyrinthine Gematria Tallies Modularly.

adamemerson commented:
This would probably be a pretty good patch to merge even as-is, though I could also put my stuff on top of it, whichever people think is better.

cbodley commented:
https://github.com/ceph/ceph/pull/12104 is the latest PR with these changes - closing this one
